modules/apps/headless/headless-form/headless-form-impl/src/main/java/com/liferay/headless/form/internal/helper/UploadFileHelper.java

```diff
@@ -0,0 +1,196 @@
+ *
+ * This library is free software; you can redistribute it and/or modify it under
+ * the terms of the GNU Lesser General Public License as published by the Free
+ * Software Foundation; either version 2.1 of the License, or (at your option)
+ * any later version.
+ *
+ * This library is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+ * FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License for more
+ * details.
+ */
+
+package com.liferay.headless.form.internal.helper;
+
+import com.liferay.document.library.kernel.service.DLAppService;
+import com.liferay.document.library.util.DLURLHelper;
+import com.liferay.dynamic.data.mapping.model.DDMFormField;
+import com.liferay.dynamic.data.mapping.model.DDMFormInstance;
+import com.liferay.dynamic.data.mapping.model.LocalizedValue;
+import com.liferay.dynamic.data.mapping.model.UnlocalizedValue;
+import com.liferay.dynamic.data.mapping.model.Value;
+import com.liferay.dynamic.data.mapping.storage.DDMFormFieldValue;
+import com.liferay.headless.form.dto.v1_0.FormDocument;
+import com.liferay.headless.form.dto.v1_0.MediaForm;
+import com.liferay.portal.kernel.exception.PortalException;
+import com.liferay.portal.kernel.json.JSONException;
+import com.liferay.portal.kernel.json.JSONFactoryUtil;
+import com.liferay.portal.kernel.json.JSONObject;
+import com.liferay.portal.kernel.repository.model.FileEntry;
+import com.liferay.portal.kernel.repository.model.FileVersion;
+import com.liferay.portal.kernel.service.ServiceContext;
+import com.liferay.portal.kernel.util.LocaleThreadLocal;
+import com.liferay.portal.vulcan.multipart.BinaryFile;
+import com.liferay.portal.vulcan.multipart.MultipartBody;
+
+import java.io.IOException;
+
+import java.util.List;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+import javax.ws.rs.BadRequestException;
+
+import org.osgi.service.component.annotations.Component;
+import org.osgi.service.component.annotations.Reference;
+
+/**
+ * @author Javier Gamarra
+ * @author Paulo Cruz
+ * @author Victor Oliveira
+ */
+@Component(immediate = true, service = UploadFileHelper.class)
+public class UploadFileHelper {
+
+	public FileEntry getFileEntry(String value) throws PortalException {
+		try {
+			JSONObject jsonObject = JSONFactoryUtil.createJSONObject(value);
+
+			return _dlAppService.getFileEntry(
+				jsonObject.getLong("fileEntryId"));
+		}
+		catch (JSONException jsone) {
+			return null;
+		}
+	}
+
+	public void linkFiles(
+		List<DDMFormField> ddmFormFields,
+		List<DDMFormFieldValue> ddmFormFieldValues) {
+
+		Stream<DDMFormField> ddmFormFieldsStream = ddmFormFields.stream();
+
+		ddmFormFieldsStream.filter(
+			formField -> Objects.equals(formField.getType(), "document_library")
+		).map(
+			field -> {
+				Stream<DDMFormFieldValue> ddmFormFieldValuesStream =
```

Comment:
this could be a flatmap right? but leave it like this, too complicated otherwise